src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -307,44 +336,118 @@ private void EventLoop()
+                }
+
+                // The native shim is responsible for ensuring this condition.
+                Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
+
+                for (int i = 0; i < numEvents; i++)
+                {
+                    IntPtr handle = _buffer[i].Data;
+                    if (handle == ShutdownHandle)
                     {
-                        throw new InternalException(err);
+                        shutdown = true;
                     }
+                    else
+                    {
+                        Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
+                        _handleToContextMap.TryGetValue(handle, out SocketAsyncContext context);
+                        if (context != null)
+                        {
+                            context.HandleEvents(_buffer[i].Events);
+                            context = null;
+                        }
+                    }
+                }
+            }
+        }
 
-                    // The native shim is responsible for ensuring this condition.
-                    Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
+        private void EventLoopBodyAio()
+        {
+            bool shutdown = false;
+            SocketAsyncContext.AsyncOperation[] batchedOperations = new SocketAsyncContext.AsyncOperation[EventBufferCount];
+            Span<Interop.Sys.IoControlBlock> ioControlBlocks = new Span<Interop.Sys.IoControlBlock>(_aioBlocks, EventBufferCount);
+            while (!shutdown)
+            {
+                int numEvents = EventBufferCount;
+                Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
+                if (err != Interop.Error.SUCCESS)
+                {
+                    throw new InternalException(err);
+                }
 
-                    for (int i = 0; i < numEvents; i++)
+                var socketEvents = new ReadOnlySpan<Interop.Sys.SocketEvent>(_buffer, numEvents);
+                int batchedCount = 0;
+                foreach (var socketEvent in socketEvents)
+                {
+                    if (socketEvent.Data == ShutdownHandle)
+                    {
+                        shutdown = true;
+                    }
+                    else if (_handleToContextMap.TryGetValue(socketEvent.Data, out SocketAsyncContext context))
                     {
-                        IntPtr handle = _buffer[i].Data;
-                        if (handle == ShutdownHandle)
+                        if ((socketEvent.Events & Interop.Sys.SocketEvents.Error) != 0)
                         {
-                            shutdown = true;
+                            // there was an error, we use the non-buffered execution path (this should be very rare)
+                            context.HandleEvents(socketEvent.Events);
```

Comment:
important: epoll can return read, write or error. If it's error, we fall back to non-AIO code path